src/datepicker/utils/types.js

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright (c) 2018-2020 Uber Technologies, Inc.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+*/
+// @flow
+
+export type DateT = moment$Moment | Date;
+
+type DateValues = DateT | string | number;
+type DateInput = ?DateValues;
+
+type Comparison = (value: DateT, comparing: DateT) => boolean;
+type DateInDateOut = (value: DateT) => DateT;
+
+type DateFuncT = (DateInput | void) => DateT;
```

Comment:
This ended up being a little bit of a compromise.  As you can see [here](https://github.com/dmtrKovalenko/date-io/blob/master/packages/core/IUtils.d.ts#L60), `date` can actually return null.  The idea is that, if you pass `null`, it will just always return `null`. 

The problem was, this was introducing all kinds of type refinement issues when functions that were previously expecting dates, were receiving this type that could be null.   I experimented with using [predicate functions](https://flow.org/en/docs/types/functions/#toc-predicate-functions) to let the flow know that if you pass anything but null into date, it will always be defined, but the fact that it was an externally defined function made that much more complicated. 

My compromise solution was just to act as if this null path didn't exist.  It seems like a little bit of an edge case situation anyway, so this seemed like the best option.

Reply:
This does add some nuance to consuming date values from props like [here](https://github.com/uber/baseweb/pull/2968/files#diff-a551f5d9fe815b88dbfdfccf4f31a6afR178) for example. Does flow raise an error if `undefined` is omitted from that line?

Without running the code, my guess is that because `DateInput` is a [maybe type](https://github.com/uber/baseweb/pull/2968/files#diff-73c5142ac939310b4e94f825bad665dbR12), the argument type provided to `DateFuncT` would expand out to `moment$Moment | Date | string | number | null | void`.  Does that seem to be the case?

My worry is that we will miss handling the null case and cause a fairly common pattern, initializing controlled state variables to `null`, to break amongst different date-io adapters